src/client/ngrx-data/entity-filter.service.ts

```diff
@@ -0,0 +1,108 @@
+import { Inject, Injectable, InjectionToken, Optional } from '@angular/core';
+import { EntityClass, getEntityName } from './interfaces';
+
+/** EntityAction Filter payload */
+export interface EntityFilter {
+  /** Name of the filter */
+  name?: string;
+  /** Search pattern for the filter to apply */
+  pattern?: any;
+}
+
+/**
+ * Filters the `entities` array argument and returns the `entities`,
+ * a memoized array of entities, or a new filtered array.
+ * NEVER mutate the `entities` array itself.
+ **/
+export type EntityFilterFn<T> = (entities: T[], pattern?: any) => T[];
+
+export interface EntityFilterDef<T> {
+  filterFn: EntityFilterFn<T>;
+  include?: string[]; // entity type names to include
+  exclude?: string[]; // entity type names to exclude
+}
+
+export interface EntityFilters { [name: string]: EntityFilterDef<any> }
+
+export const ENTITY_FILTERS = new InjectionToken<EntityFilters>('ENTITY_FILTERS');
+
+/** EntityFilter function: match pattern in the entity name. */
+export function GenericNameFilterFn<T>(entities: T[], pattern: string) {
```

Comment:
This always matches the `name`. Do we really want to assume an entity has a specific property?

Reply:
Actually, if the entity has no `name` property, the entity is excluded. I could have gone the other way and included it. Or thrown.

The default is whatever filter matches the unnamed filter, ''`.  Our out-of-the-box filter function is the `GenericNameFilterFn` but it could be anything we want to ship. I just wanted us to ship with _something_.